invenio_matcher/matcherext/engines/es.py

```diff
@@ -0,0 +1,203 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Invenio Matcher Elasticsearch engine."""
+
+from flask import current_app
+
+from invenio.base.wrappers import lazy_import
+
+from invenio_matcher.errors import InvalidCall
+from invenio_matcher.errors import NoQueryDefined
+from invenio_matcher.models import MatchResult
+
+import six
+
+from werkzeug.utils import import_string
+
+
+ES = lazy_import('invenio.ext.es.es')
+Record = lazy_import('invenio_records.api.Record')
+
+
+def exact(**kwargs):
+    """Build an exact query and send it to Elasticsearch."""
+    return _query(_build_exact_query, **kwargs)
+
+
+def fuzzy(**kwargs):
+    """Build a fuzzy query and send it to Elasticsearch."""
+    return _query(_build_fuzzy_query, **kwargs)
+
+
+def free(**kwargs):
+    """Build a free query and send it to Elasticsearch."""
+    return _query(_build_free_query, **kwargs)
+
+
+def queries(**kwargs):
+    """Return queries defined for the given index and doc_type."""
+    try:
```

Comment:
Wait. `engine.py` can not know that it should call `exact(index, doc_type, **kwargs)`, since it's an implementation detail specific to Elasticsearch, so it has to call `exact(**kwargs)`. What I can do is move the logic that checks for required keyword arguments in a decorator, so that the code reads nicer. Would that be ok?